src/Chainweb/CutDB.hs

```diff
@@ -251,6 +253,26 @@ _cutStm = readTVar . _cutDbCut
 cutStm :: Getter (CutDb cas) (STM Cut)
 cutStm = to _cutStm
 
+-- | A common idiom to spin while waiting for a guaranteed new `Cut`, different
+-- from the given one.
+--
+awaitNewCut :: CutDb cas -> Cut -> IO Cut
+awaitNewCut cdb c = atomically $ do
+    c' <- _cutStm cdb
+    when (c' == c) retry
+    return c'
+
+-- | As in `awaitNewCut`, but only updates when the specified `ChainId` has
+-- grown.
+--
+awaitNewCutByChainId :: CutDb cas -> ChainId -> Cut -> IO Cut
+awaitNewCutByChainId cdb cid c = atomically $ do
+    c' <- _cutStm cdb
+    let !b0 = HM.lookup cid $ _cutMap c
+        !b1 = HM.lookup cid $ _cutMap c'
+    when (b1 == b0) retry
+    return c'
```

Comment:
This function is the vehicle by which we do non-wasteful preempting for both local and remote mining. New `Cut`s come in all the time, but a particular mining process only wants to stop if **the chain they're currently working on** has changed. Otherwise, their work shouldn't be considered stale.

This will increase mining efficiency on our current network dramatically.